crates/viewer/re_space_view_dataframe/src/expanded_rows.rs

```diff
@@ -0,0 +1,136 @@
+pub(crate) struct ExpandedRowsCache {
+    /// Maps "table row number" to "additional expanded rows".
+    ///
+    /// When expanded, the base space is still used for the summary, which the additional space is
+    /// used for instances.
+    expanded_rows: BTreeMap<u64, u64>,
+
+    /// ID used to invalidate the cache.
+    valid_for: egui::Id,
+}
+
+impl Default for ExpandedRowsCache {
+    fn default() -> Self {
+        Self {
+            expanded_rows: BTreeMap::default(),
+            valid_for: egui::Id::new(""),
+        }
+    }
+}
+
+impl ExpandedRowsCache {
+    /// This sets the query used for cache invalidation.
+    ///
+    /// If the query doesn't match the cached one, the state will be reset.
+    fn validate_id(&mut self, id: egui::Id) {
+        if id != self.valid_for {
+            self.valid_for = id;
+            self.expanded_rows = BTreeMap::default();
+        }
+    }
+}
+
+/// Helper to keep track of row expansion.
+///
+/// This is a short-lived struct to be created every frame. The persistent state is stored in
+/// [`ExpandedRowsCache`].
+///
+/// Uses egui's animation support to animate the row expansion/contraction. For this to work:
+/// - When collapsed, the row entry must be set to 0 instead of being removed. Otherwise, it will no
+///   longer be "seen" by the animation code. Technically, it could be removed _after_ the
+///   animation completes, but it's not worth the complexity.
+/// - When the row is first expanded, for the animation to work, it must be immediately seeded to 0
+///   for the animation to have a starting point.
+pub(crate) struct ExpandedRows<'a> {
+    /// Base row height.
+    row_height: f32,
+
+    /// Cache containing the row expanded-ness.
+    cache: &'a mut ExpandedRowsCache,
+
+    /// [`egui::Context`] used to animate the row expansion.
+    egui_ctx: egui::Context,
+
+    /// [`egui::Id`] used to store the animation state.
+    id: egui::Id,
+}
+
+impl<'a> ExpandedRows<'a> {
+    /// Create a new [`ExpandedRows`] instance.
+    ///
+    /// `egui_ctx` is used to animate the row expansion
+    /// `id` is used to store the animation state and invalidate the cache, make it persistent and
+    /// unique
+    pub(crate) fn new(
+        egui_ctx: egui::Context,
+        id: egui::Id,
+        cache: &'a mut ExpandedRowsCache,
+        row_height: f32,
+    ) -> Self {
+        // (in-)validate the cache
+        cache.validate_id(id);
+
+        Self {
+            row_height,
+            cache,
+            egui_ctx,
+            id,
+        }
+    }
+
+    /// Implementation for [`egui_table::TableDelegate::row_top_offset`].
+    pub(crate) fn row_top_offset(&self, row_nr: u64) -> f32 {
+        self.cache
+            .expanded_rows
+            .range(0..row_nr)
+            .map(|(expanded_row_nr, expanded)| {
+                self.egui_ctx.animate_value_with_time(
+                    self.row_id(*expanded_row_nr),
+                    *expanded as f32 * self.row_height,
```

Comment:
Reading the code on GitHub, I can't tell if `expanded` is a `bool` or an integer.

If it is a bool, it should be called `is_expanded`, but then this is buggy (an expanded row is only twice as big as an unexpanded row).

So I'm guessing this is an integer:

```suggestion
            .map(|(expanded_row_nr, expanded_row_count)| {
                self.egui_ctx.animate_value_with_time(
                    self.row_id(*expanded_row_nr),
                    *expanded_row_count as f32 * self.row_height,
```

Reply:
Renamed this and a bunch of other things in terms of "additional lines"